Added a one-off job for math expression input validation.

iamprayush opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

~~1. This PR fixes or fixes part of #[fill_in_number_here].~~
2. This PR does the following:
Adds a one-off job that validates all explorations that use the MathExpressionInput interaction to check how many inputs are valid according to the new validations and also to classify them between equations and expressions. This is a prerequisite for the upgradation of the explorations to the new math interactions. 

This PR also removes the old one-off job that was added to check how many inputs are equations. That functionality is a subset of this one-off job which is why the old one is now obsolete.

The following screenshots show an example run of the job that was carried out locally:
Sample exploration inputs:
![inputs](https://user-images.githubusercontent.com/35144226/85206876-c1de3480-b342-11ea-9778-ec125f99764a.png)

Job output:
![job_output](https://user-images.githubusercontent.com/35144226/85206883-c6a2e880-b342-11ea-9ab9-8ed429f18fa2.png)


## Essential Checklist

- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## PR Pointers

- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays
- Never force push. If you do, your PR will be closed.

iamprayush committed:
Added a one-off job for validation and removed the old one.

aks681 commented on core/domain/exp_jobs_one_off.py:
```diff
@@ -116,10 +128,12 @@ def reduce(key, values):
         yield (key, values)
 
 
-class MathExpressionInputInteractionOneOffJob(jobs.BaseMapReduceOneOffJobManager):  # pylint: disable=line-too-long
```
Is this job not required any more?

iamprayush replied:
The old job just checked which interactions have equations as inputs, so this job covers that functionality which is why we can remove the old one.

iamprayush commented on core/domain/exp_jobs_one_off.py:
```diff
@@ -34,8 +36,18 @@
+import schema_utils
 import utils
 
+CURR_DIR = os.path.abspath(os.getcwd())
```
This is done so that the path to the third-party lib can be resolved. Also, the linter disables are necessary because I couldn't add the path insert code above the imports, that would also cause the linter checks to fail.

seanlip replied:
I don't understand why all this is needed, especially since we seem to be able to import other third-party libs without a problem (see e.g. mutagen). Why is this case different and why does it need so much special-case setup? In general, please make sure to follow existing patterns in the codebase.

iamprayush replied:
The other libs also require such explicit addition of paths. The lib you've mentioned has done this in multiple places that it's being used in. Since this is the only place pylatexenc is being used in, the path to it has been added here. Also, this sort of thing is being done in other places as well, for instance [here](https://github.com/oppia/oppia/blob/640cf5aff24da4df8e549f4d79b9b704ba74f213/scripts/linters/js_ts_linter.py#L35), the path to the esprima library has been explicitly added just the way I've added it here which is why I thought this was the norm.

seanlip replied:
The lib I mentioned appears to be used in exactly one place (core/controllers/voice_artist.py) and, as far as I can tell, does not require any addition of paths or usage of linter disables. Could you please clarify what you are looking at?

Stuff in the scripts/ folder has different rules because it is run using cmdline, so don't take that as a basis for what you should do with third-party libs in the main codebase.

iamprayush replied:
I've fixed this now.

seanlip replied:
Thanks. Could you please explain the earlier sentence about "The lib you've mentioned has done this in multiple places that it's being used in." -- what happened there?

iamprayush replied:
Ah, I saw the path being inserted in multiple files like `appengine_config.py`, `gae_suite.py`, `run_backend_tests.py`, and `pre_commit_linter.py`. I thought if I were to just add the path in this file, I wouldn't have to add it to multiple places like that. But I suppose that is more consistent, which is why I've added it to those files instead now.

aks681 reviewed (approved):
Lgtm

aks681 commented:
@seanlip @DubeySandeep @vojtechjelinek PTAL.

seanlip commented on core/domain/exp_jobs_one_off.py:
```diff
@@ -116,10 +128,12 @@ def reduce(key, values):
+    """Job that produces a list of explorations that use the MathExpressionInput
+    along with the validity and type(expression/equation) of the inputs present
+    in the exploration.
+    This validation is done by the 'is_valid_math_expression' or the
```
Always add a single new line between paragraphs.

iamprayush replied:
Done.

seanlip commented on core/domain/exp_jobs_one_off.py:
```diff
@@ -128,19 +142,29 @@ def entity_classes_to_map_over(cls):
+            'is_valid_math_expression')
+        is_valid_math_equation = schema_utils.get_validator(
+            'is_valid_math_equation')
+        ltt = latex2text.LatexNodes2Text()
```
I thought we were planning to validate the asciimath, not the latex. 

Also, is this library meant to be a permanent addition? What is the precise definition of the format of the text output of latex2text? Is it asciimath? If so, could you point me to the docs which state that? If not, then wouldn't this mean that there is a mismatch?

iamprayush replied:
> I thought we were planning to validate the asciimath, not the latex.

Currently the input is stored in the form of latex which is why we need to convert it before validating it. Also, when we perform the actual upgrade, we would have to replace the latex string with the text format to maintain consistency.

> Also, is this library meant to be a permanent addition?

No. Once the upgradation of the old explorations has been done successfully, we can remove this from the codebase.

> What is the precise definition of the format of the text output of latex2text? Is it asciimath? If so, could you point me to the docs which state that? If not, then wouldn't this mean that there is a mismatch?

The [documentation](https://pylatexenc.readthedocs.io/en/latest/latex2text/) doesn't mention the exact format of the text output with the exact grammar; but I've played around with this locally and it seems to convert the latex syntax into human readable string. So things like `\frac{x}{y}` turns into `x/y` and so on. Note that there is no tight requirement for having asciimath as long as it is in a human-readable form. The reason for choosing asciimath was that it was the closest to a human-readable form that guppy outputted in,

seanlip replied:
I'm afraid I don't follow. A MathExpression has an ascii property and a latex property (look at the definition of the object in extensions/objects). If I understand correctly, the former is supposed to be in asciimath. So why don't you just validate that and avoid latex conversion altogether?

iamprayush replied:
Ah, I understand what the confusion is. The thing is, the expected answer in this interaction is indeed in the form of an object that contains both, latex as well as asciimath. But, this is only for the learner's answer. The creator's input is in Unicode because while creating an exploration with this interaction, the creator gets a normal text input box instead of the guppy editor. And obviously, the text input cannot output both latex as well as asciimath. Here, we are just expecting the creator to enter a correctly formatted latex string without any validations to check if the string entered is correctly formatted. I'm not sure what the thought process was behind not using the guppy editor in the creator's view as well, but because of this, we don't have the asciimath format of the creator's input, which is what this job validates. Does that clear things up?

seanlip replied:
Yes, this makes sense, thanks for the clear explanation. Just to check -- the new interactions will be using guppy input, and the stored format will be math expression input with both ascii and latex fields, right? (And we'll be migrating from the old unicodestring to the new mathexpression format?) If so, all that sounds good.

(I'm not sure why the unicode editor was used originally, but I agree that it's definitely suboptimal.)

iamprayush replied:
Yes, the new interactions will all use guppy for the learner and the creator's views. The answer will be stored as `AlgebraicExpression`, `MathEquation`, or `NumericExpression`. But only asciimath format from the guppy editor will be stored since we don't need latex. And yes, all the old latex input strings will be converted to ascii and will be stored in the new format.

seanlip replied:
OK, that sounds good, thanks.

Just to check, would it be possible to convert from asciimath to latex, if we ever wanted to do so? (If so, how?)

iamprayush replied:
Yes, pylatexenc also has a reversed functionality to convert text to latex. Also, if we ever need latex in the frontend, nerdamer has functionality to convert the expression/equation to latex.

seanlip replied:
OK, sounds good. Thanks!

seanlip commented on core/domain/exp_jobs_one_off.py:
```diff
@@ -128,19 +142,29 @@ def entity_classes_to_map_over(cls):
+                            yield (
+                                item.id,
+                                u'%s: %s -- %s' % (
+                                    state_name, rule_input, validity))
```
This seems a bit off. If what you want to do is view the problematic cases rather than list all formulae, then perhaps your first element (the key for the mapreduce) should be by validity and the second element should contain all the details. For the valid cases you probably can just output the first N (100-1000?) rule_input strings for eyeballing, but for the invalid cases you'll want to output all of them with full details.

iamprayush replied:
That makes sense. I've set the limit for valid explorations to 200.

seanlip reviewed (changes requested):
Thanks, @iamprayush -- took a pass.

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into math-exp-one-off

iamprayush committed:
Added a threshold for displaying valid explorations.

iamprayush committed:
Fixed linter errors.

iamprayush committed:
Fixed linter errors.

seanlip commented on core/domain/exp_jobs_one_off.py:
```diff
@@ -116,34 +128,54 @@ def reduce(key, values):
+                            validity = 'Invalid'
+                            if is_valid_math_expression(rule_input):
+                                validity = 'Valid Expression'
+                                MathExpressionValidationOneOffJob.valid_exps_seen += 1  #pylint: disable=line-too-long
```
Why are you storing this on the class? Just declare local vars.

Also I feel like this completely misses the point of mapreduce jobs. Please read up a bit more on how these work and look at other examples in the codebase. Any combination of data between individual map threads should be done in reduce(). The whole point of MR is so that things can run on a distributed system so don't assume that the same running instance is processing all of the models. In the map() phase you'll still need to yield everything, let reduce() sort out the combination / filtering / etc.

iamprayush replied:
Oh, I got confused. I've localized the variable now, thanks!

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into math-exp-one-off

iamprayush committed:
Localized valid exp counter and fixed insertion of path for pylatexenc.

seanlip commented on appengine_config.py:
```diff
@@ -89,6 +89,10 @@ def save(self):
         raise Exception('Invalid path for oppia_tools library: %s' % pil_path)
     sys.path.insert(0, pil_path)
 
+OPPIA_TOOLS = [
```
Why are you putting this here? Put it in the usual place, similar to all other third-party libs.

OPPIA_TOOLS stuff does not get uploaded to production.

iamprayush replied:
I've removed this.

seanlip commented on core/domain/exp_jobs_one_off.py:
```diff
@@ -33,14 +33,16 @@
 gae_image_services = models.Registry.import_gae_image_services()
 
-SEPARATORS = ['<', '>', '=']
+MATH_VALID_EXP_LIMIT = 200
```
I don't know what this means. Please pick a clearer name.

iamprayush replied:
I've updated the var name and added a comment.

seanlip commented on core/domain/exp_jobs_one_off.py:
```diff
@@ -33,14 +33,16 @@
 from core.domain import rights_manager
 from core.platform import models
 import feconf
+from pylatexenc import latex2text
```
Normally, third-party stuff comes after our internal stuff -- see the mutagen example again. Does pylint require you to place this here? If so, that's a bit odd and you should talk to @Hudda.

iamprayush replied:
@Hudda I'm getting an `Imports are incorrectly sorted` error if I import this third-party lib after the internal imports. Could you PTAL? Thanks!

Edit: The linter couldn't recognize this as a third party lib -- I had to add this lib to `.isort.cfg` and it's working as expected.

seanlip commented on core/domain/exp_jobs_one_off.py:
```diff
@@ -116,10 +118,13 @@ def reduce(key, values):
-    identify and prevent usage of equation/inequalities.
+class MathExpressionValidationOneOffJob(jobs.BaseMapReduceOneOffJobManager):
+    """Job that produces a list of explorations that use the MathExpressionInput
+    along with the validity and type(expression/equation) of the inputs present
```
Add space before (

iamprayush replied:
Done.

seanlip commented on core/domain/exp_jobs_one_off.py:
```diff
@@ -128,19 +133,32 @@ def entity_classes_to_map_over(cls):
+                            elif is_valid_math_equation(rule_input):
+                                validity = 'Valid Equation'
+                                valid_exps_seen += 1
+                            if valid_exps_seen <= MATH_VALID_EXP_LIMIT:
```
You aren't going to reach this limit in a single exploration. The idea was to cap the output, so the limit should be applied in the reduce() step.

For invalid stuff, presumably you will want to include more debug info so that the issues can be fixed manually if necessary. E.g. the exp ID and some info about where the problematic expression was found.

iamprayush replied:
> the limit should be applied in the reduce() step.

@seanlip Just to confirm, we would then have to use a class variable like I'd done before, right? If I'm not mistaken, static methods (map and reduce) won't be able to access instance variables (if I were to initialize it in the `__init__` method).

Edit: I just realized this can be done without the static variable as well. Please ignore the above question.

seanlip commented:
Please see additional comments [here](https://github.com/oppia/oppia/pull/9597#discussion_r443357032) and [here](https://github.com/oppia/oppia/pull/9597#discussion_r443356456), too.

iamprayush committed:
Modified valid expression counter and other minor nits.

seanlip commented on core/domain/exp_jobs_one_off.py:
```diff
@@ -34,13 +34,23 @@
 gae_image_services = models.Registry.import_gae_image_services()
 
-SEPARATORS = ['<', '>', '=']
+# This threshold puts a cap on the number of valid explorations that can be
```
expressions, not explorations

iamprayush replied:
Done.

seanlip commented on core/domain/exp_jobs_one_off.py:
```diff
@@ -34,13 +34,23 @@
-SEPARATORS = ['<', '>', '=']
+# This threshold puts a cap on the number of valid explorations that can be
+# yielded by the MathExpressionValidationOneOffJob.
+VALID_MATH_EXP_YIELD_LIMIT = 200
```
Do not abbreviate -- there is ambiguity about exploration vs expression

iamprayush replied:
Done.

seanlip commented on core/domain/exp_jobs_one_off.py:
```diff
@@ -34,13 +34,23 @@
+# This threshold puts a cap on the number of valid explorations that can be
+# yielded by the MathExpressionValidationOneOffJob.
+VALID_MATH_EXP_YIELD_LIMIT = 200
+UNICODE_TO_TEXT = {
```
Put this (and probably the constant above) at the class level. It doesn't need to be exposed to the entire module because it's very specific to one particular job (but its naming makes it sound generic).

iamprayush replied:
Done.

seanlip reviewed (changes requested):
Hi @iamprayush -- looks good! Just minor notes, but I think you can now file the job request and get all the other reviewers' approval.

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into math-exp-one-off

iamprayush committed:
Fixed minor nits.

seanlip commented on scripts/run_backend_tests.py:
```diff
@@ -73,6 +73,8 @@
     os.path.join(common.OPPIA_TOOLS_DIR, 'psutil-%s' % common.PSUTIL_VERSION),
     os.path.join(
         common.OPPIA_TOOLS_DIR, 'PyGithub-%s' % common.PYGITHUB_VERSION),
+    os.path.join(
```
I think something is wrong with your setup here. Why are you still installing this in oppia_tools?

Please fix this -- I tried running your PR on my local machine and it gave a 500 server error. You can test by checking out a copy of Oppia in a second dir on your local machine and performing a full install from scratch -- make sure the start script results in the page loading without any issues.

iamprayush replied:
> I think something is wrong with your setup here. Why are you still installing this in oppia_tools?

I saw that the other libs which are installed using pip are installed in oppia_tools. I'll change it and add the link in `manifest.json` instead.

> I tried running your PR on my local machine and it gave a 500 server error.

It was giving me a 500 error as well which is why I had to insert the path in a bunch of places. It works for me now, still, I'll double-check it in a fresh copy.

seanlip replied:
Remember: oppia_tools is for stuff that is not uploaded to prod. It's for developer scripts, and stuff that doesn't already exist in App Engine. If you want something to be in prod, then it needs to go in third_party.

Is the library not on github? You could install from there instead perhaps.

iamprayush replied:
I've moved this from oppia_tools to third_party_libs and I've tested it locally on a fresh copy of the repo.

seanlip commented on core/tests/gae_suite.py:
```diff
@@ -45,6 +45,7 @@
     os.path.join(OPPIA_TOOLS_DIR, 'Pillow-6.2.2'),
     os.path.join(OPPIA_TOOLS_DIR, 'psutil-5.7.0'),
     os.path.join(OPPIA_TOOLS_DIR, 'PyGithub-1.45'),
+    os.path.join(OPPIA_TOOLS_DIR, 'pylatexenc-2.5'),
```
Wrong dir. I thought I mentioned this in a previous review?

iamprayush replied:
Changed.

seanlip commented on core/domain/exp_jobs_one_off.py:
```diff
@@ -160,7 +161,7 @@ def map(item):
                             # they need to be replaced with their corresponding
                             # text values before performing validation.
-                            for unicode_char, text in UNICODE_TO_TEXT.items():
+                            for unicode_char, text in MathExpressionValidationOneOffJob.UNICODE_TO_TEXT.items(): #pylint: disable=line-too-long
```
Avoid pylint pragmas. You can define a local var with a shorter name, for example.

In general please try to really avoid pylint pragmas in your code, going forward. The first thing the reviewer will do is to ask you why you need them in the first place.

iamprayush replied:
Understood. I've changed it now.

iamprayush committed:
Changed installation path of pylatexenc to third_party_libs.

iamprayush committed:
Added exploration id to the job output.

seanlip commented on core/domain/exp_jobs_one_off.py:
```diff
@@ -146,6 +147,7 @@ def map(item):
         is_valid_math_equation = schema_utils.get_validator(
             'is_valid_math_equation')
         ltt = latex2text.LatexNodes2Text()
+        mevooj = MathExpressionValidationOneOffJob
```
Don't give it a made-up name. Use something semantic like "cls" or current_class. Or, perhaps better:

```
unicode_to_text_mapping = (
    MathExpressionValidationOneOffJob.UNICODE_TO_TEXT)
```

iamprayush replied:
Done.

seanlip commented on core/domain/exp_jobs_one_off.py:
```diff
@@ -172,12 +175,14 @@ def map(item):
+        mevooj = MathExpressionValidationOneOffJob
         if key.startswith('Valid'):
-            yield (key, values[:VALID_MATH_EXP_YIELD_LIMIT])
+            yield (key, values[:mevooj.VALID_MATH_INPUTS_YIELD_LIMIT])
```
Ditto here and elsewhere.

iamprayush replied:
Done.

seanlip commented:
Also, I did a preliminary test run and we can't merge this yet. The following types of errors are showing up:

```
Traceback (most recent call last): (/third_party/gae-mapreduce-1.9.22.0/mapreduce/handlers.py:517)
  File "/third_party/gae-mapreduce-1.9.22.0/mapreduce/handlers.py", line 503, in handle
    tstate.input_reader, shard_state, tstate, ctx)
  File "/third_party/gae-mapreduce-1.9.22.0/mapreduce/handlers.py", line 593, in _process_inputs
    entity, input_reader, ctx, tstate):
  File "/third_party/gae-mapreduce-1.9.22.0/mapreduce/handlers.py", line 644, in _process_datum
    output_writer.write(output)
  File "/third_party/gae-mapreduce-1.9.22.0/mapreduce/output_writers.py", line 1151, in write
    value = str(data[1])
UnicodeEncodeError: 'ascii' codec can't encode character u'\xe9' in position 32: ordinal not in range(128)


Traceback (most recent call last): (/third_party/gae-mapreduce-1.9.22.0/mapreduce/handlers.py:517)
  File "/third_party/gae-mapreduce-1.9.22.0/mapreduce/handlers.py", line 503, in handle
    tstate.input_reader, shard_state, tstate, ctx)
  File "/third_party/gae-mapreduce-1.9.22.0/mapreduce/handlers.py", line 593, in _process_inputs
    entity, input_reader, ctx, tstate):
  File "/third_party/gae-mapreduce-1.9.22.0/mapreduce/handlers.py", line 644, in _process_datum
    output_writer.write(output)
  File "/third_party/gae-mapreduce-1.9.22.0/mapreduce/output_writers.py", line 1151, in write
    value = str(data[1])
UnicodeEncodeError: 'ascii' codec can't encode character u'\xf3' in position 20: ordinal not in range(128)


Traceback (most recent call last): (/third_party/gae-mapreduce-1.9.22.0/mapreduce/handlers.py:517)
  File "/third_party/gae-mapreduce-1.9.22.0/mapreduce/handlers.py", line 503, in handle
    tstate.input_reader, shard_state, tstate, ctx)
  File "/third_party/gae-mapreduce-1.9.22.0/mapreduce/handlers.py", line 593, in _process_inputs
    entity, input_reader, ctx, tstate):
  File "/third_party/gae-mapreduce-1.9.22.0/mapreduce/handlers.py", line 644, in _process_datum
    output_writer.write(output)
  File "/third_party/gae-mapreduce-1.9.22.0/mapreduce/output_writers.py", line 1151, in write
    value = str(data[1])
UnicodeEncodeError: 'ascii' codec can't encode characters in position 13-24: ordinal not in range(128)
```

iamprayush committed:
Added code to handle out of range characters.

iamprayush commented:
@seanlip That seems to be caused due to accented characters being present in the input. While yielding, the `rule_input` is decoded but since these characters are out of range, the error gets thrown. I've added a fix for this wherein such characters get replaced by their ordinal values and get added to the invalid inputs. I've also added a test case for this.

seanlip commented on core/domain/exp_jobs_one_off.py:
```diff
@@ -116,35 +118,85 @@ def reduce(key, values):
+                            # special characters like sqrt and pi, which is why
+                            # they need to be replaced with their corresponding
+                            # text values before performing validation.
+                            for unicode_char, text in (
```
The typical way to handle this sort of thing is to use .encode('utf-8'). You'll see this in existing tests. Why don't you do something like that instead, rather than all this special-casing?

iamprayush replied:
The purpose of doing this is to replace special characters' unicode values like u221a for sqrt with their string representations. So for instance, `\sqrt{2}` will get converted to `\u221a(2)` by pylatexenc. So, after replacing with the corresponding text value, we'll get `sqrt(2)` which is exactly what we want.

seanlip replied:
Oh, I see, you need it for is_valid_math_expression().

OK then can you apply the .encode() thing instead of lines 171-179 below? That is how we typically handle "out of range" chars.

iamprayush replied:
@seanlip It's not only for the validation functions. Even while yielding the result, it is required to not have characters with ordinal values greater than 127. Using `.encode()` gives the following error:

```
File "/home/prayush/Desktop/opensource/oppia/core/domain/exp_jobs_one_off.py", line 192, in map
    item.id, state_name.encode('utf-8'), rule_input.encode('utf-8')))
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 0: ordinal not in range(128)
```

seanlip replied:
Can you check where the encode() error is happening -- is it in the state name or the rule input?

What is an example of an input that causes this -- can you walk me through a trace of how it goes through the various stages of your map()?

iamprayush replied:
It's happening due to the rule input. I'll explain this through the example I've added to the tests. Let's assume the rule input is `âéîôü` (contains characters that have ordinal values > 127). If these characters are from the ones that we are expecting, like the square root sign or the multiplication dot, then we simply replace them with the corresponding strings (`√` -> `sqrt`) and everything is fine. But, if the input contains characters that we are not expecting, they wouldn't get replaced and these "out of range" characters would still be present in the rule_input string. So, when we try to perform some operation on this string (during validation) or when they are yielded by the map function, the string gets decoded but due to these "out of range" characters, a UnicodeDecodeError is thrown. To avoid this, what I've done is, I've replaced all characters that have ordinal values > 127 with their corresponding ordinal values. This is essentially the same as replacing square root's character with `sqrt`, but since we can't special-case _all_ possible characters, we just replace them with the ordinal value. So, the string `âéîôü` gets converted to `226 233 238 244 252` which does not contain characters that are out of range and everything works smoothly. Also, note that that particular input is going to be marked as invalid anyways(since it contains unexpected characters) so we can modify it without affecting the validation process.

seanlip replied:
Tried it out. Still no good, I'm afraid -- the same errors still arise.

I would encourage you to take a closer look at encode('utf-8'). We've run into unicode issues before and they are often due to the MapReduce pipeline handling strings rather than unicode. It seems surprising to me that you don't need them for this job and you need to roll your own custom implementation. You might want to look at the patterns in other jobs and compare and contrast them to what you are doing here.

In general, try to avoid doing custom handcrafted things. Figure out what the output of pylatexenc is (str/unicode/etc.) and encode/decode it as needed using a standard method so that you can have guarantees on the types/encoding of your strings. Feel free to add assert statements/checks within the function to verify those guarantees -- if any of those assertions errors out on a production run, I will let you know.

Also, test thoroughly locally with unicode strings in various places. The error may not occur where you think it does. I suspect it's occurring because you're passing unicode chars to the MR framework somehow when it's expecting a string.

iamprayush replied:
Sorry about this, @seanlip. I have removed the customized replacements for out of range characters and used `.encode('utf-8')` instead. I think the errors are occurring because the yielded output is in unicode and as you mentioned, they need to be strings. I've tested the latest changes locally and they do work -- please let me know if there's a specific way that I can test them to ensure that they would work on production.

I think the reason my previous method did not work is probably because some of the unicode characters might still be present in the string that is being yielded. So, the latest changes ensure that there would be no such problems.

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into math-exp-one-off

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into math-exp-one-off

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into math-exp-one-off

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into math-exp-one-off

iamprayush committed:
Fixed encoding error.

DubeySandeep reviewed (approved):
LGTM for the code owner files!

seanlip reviewed (approved):
Thanks, @iamprayush. This PR is approved for merge.

seanlip commented:
(Though you'll need to get the CI checks to pass, first.)

seanlip commented:
@vojtechjelinek @Hudda @nithusha21 @bansalnitish PTAL as codeowners. Thanks!

vojtechjelinek reviewed (approved):
LGTM!

Hudda reviewed (approved):
LGTM as code owner.

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into math-exp-one-off

iamprayush committed:
Added modifications for trig functions and extended unicode to text mapping.

iamprayush commented:
@bansalnitish PTAL as code-owner, thanks!

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into math-exp-one-off

iamprayush committed:
Fixed linter errors.

bansalnitish reviewed (approved):
LGTM!

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into math-exp-one-off
